### PR TITLE
Use Roslyn to create ref/ assemblies

### PR DESCRIPTION
**Pull request by dougbu**
- #14801 

Suggest reviewing the commits individually and mostly ignoring the first ("Remove all ref/ projects"). 

This set of changes
- reduces the complexity of our build, mostly because it removes all ref/ projects
- avoids any need for manual changes when APIs change
- increases the targeting pack size slightly e.g. `Microsoft.AspNetCore.App.Ref.*.nupkg` was about 1.5 MB before this change and 2 MB after
   - most of this comes from including more information in the assemblies e.g. `internal` symbols when IVT is used
- makes it difficult to display public API changes -- until we decide on the baselines to compare against and implement that comparison

### CODE REVIEW

**Comment by Tratcher** [comment]
@dougbu you are my new favorite person 😁. 

Can /docs/ReferenceAssemblies.md be removed?

Fixes https://github.com/dotnet/aspnetcore-internal/issues/2693

**Comment by dougbu** [comment]
> Can /docs/ReferenceAssemblies.md be removed?

Yes 😺

File: eng/SharedFramework.External.props
Comment (wtgodbe) [inline eng/SharedFramework.External.props]: Why the exclamation points - was this a comment to yourself that you removed later?

File: Directory.Build.props
Comment (wtgodbe) [inline Directory.Build.props:129]: nit - change to "This is used to..." to avoid devs accidentally interpreting this as "In the past, we only produced..."

File: eng/SharedFramework.External.props
Comment (wtgodbe) [inline eng/SharedFramework.External.props]: We can probably just move this `PropertyGroup` directly into the ref-pack's project, now that it's only used there

File: eng/targets/ResolveReferences.targets
Comment (wtgodbe) [inline eng/targets/ResolveReferences.targets:81]: Why move the conditions to a separate line for other items/properties, but keep the condition on the same line here?

File: eng/targets/ResolveReferences.targets
Comment (wtgodbe) [inline eng/targets/ResolveReferences.targets:243]: How/where do we compute `HasReferenceAssembly` now?

File: eng/Dependencies.props
Comment (wtgodbe) [inline eng/Dependencies.props:197]: This scares me a _little_ bit - how strong is the guarantee that package version names will always follow this scheme? I believe darc/maestro require it (CC @mmitche), but it seems like we could easily break it for manual version updates like xunit.

File: eng/Dependencies.props
Comment (wtgodbe) [inline eng/Dependencies.props:201]: Add a comment explaining why these 2 need to be done manually?

File: eng/Versions.props
Comment (wtgodbe) [inline eng/Versions.props:167]: Why did you remove `AND '$(MSBuildProjectName)' == 'Microsoft.AspNetCore.App.Ref'` from the condition? We still want to reference the latest everywhere _except_ the ref pack

File: Directory.Build.targets
Comment (wtgodbe) [inline Directory.Build.targets:109]: Ah, here it is

File: src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj
Comment (wtgodbe) [inline src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:128]: nit - make capitalization consistent with other references to `m.nc.app`

File: docs/ReferenceResolution.md
Comment (wtgodbe) [inline docs/ReferenceResolution.md:76]: aspnetcore -> aspnetcore-tooling

File: docs/ReferenceResolution.md
Comment (wtgodbe) [inline docs/ReferenceResolution.md:77]: Same here

File: docs/ReferenceResolution.md
Comment (wtgodbe) [reply docs/ReferenceResolution.md:76]: Actually, we don't depend on tooling anymore - I think we can get rid of all CPDs in this repo now that the only repo below us is dotnet/runtime (@mmitche please confirm)

**Comment by wtgodbe** [comment]
I must have missed it - where/how do we actually hook in to the roslyn `ref` project generation?

File: eng/Versions.props
Comment (dougbu) [reply eng/Versions.props:167]: Because the condition was incorrect in 'master'. The 'release/3.1' condition is closer to what's needed: `$(MSBuildProjectName)' == 'Microsoft.AspNetCore.App.Ref' OR '$(IsReferenceAssemblyProject)' == 'true'`. But, that's incomplete because we split the ref/ and implementation versions in that branch.

I could conditionally compile implementation projects against the assemblies in these packages and test against the latest versions. But, I'm not sure that's worthwhile. Thoughts?

File: docs/ReferenceResolution.md
Comment (dougbu) [reply docs/ReferenceResolution.md:76]: No; runtime bits flow into aspnetcore and from there into aspnetcore-tooling. This part of the document describes the CPD setup needed to keep that coherent.

File: docs/ReferenceResolution.md
Comment (dougbu) [reply docs/ReferenceResolution.md:77]: As above

File: eng/targets/ResolveReferences.targets
Comment (dougbu) [reply eng/targets/ResolveReferences.targets:81]: Mostly depended on line length and where my cursor was when I hit **Enter** 😈 I don't particularly care whether the `Include` or the `Condition` appears first or second. But, if you do, lemme know.

File: eng/SharedFramework.External.props
Comment (dougbu) [reply eng/SharedFramework.External.props]: Yes, I removed this later.

File: eng/SharedFramework.External.props
Comment (dougbu) [reply eng/SharedFramework.External.props]: It's used everywhere now.

File: eng/Dependencies.props
Comment (dougbu) [reply eng/Dependencies.props:197]: If we break it, `@(Reference)` items fail loudly.

**Comment by dougbu** [comment]
> where/how do we actually hook in to the roslyn `ref` project generation?

It's a single property setting [here](https://github.com/dotnet/aspnetcore/pull/23403/files#diff-56f8482aeb786dd105fbf93dedf2615fR112).

File: eng/targets/ResolveReferences.targets
Comment (JunTaoLuo) [inline eng/targets/ResolveReferences.targets:245]: Why isn't this an error. Are there actually cases of this being valid? We could add an option to opt out of the error but by default I feel this is wrong enough to be an error.

File: eng/targets/ResolveReferences.targets
Comment (JunTaoLuo) [inline eng/targets/ResolveReferences.targets:247]: Ditto

File: eng/Versions.props
Comment (JunTaoLuo) [reply eng/Versions.props:167]: Why would this not be worthwhile?

File: docs/ReferenceResolution.md
Comment (JunTaoLuo) [reply docs/ReferenceResolution.md:76]: Yes language here is correct.

File: eng/Versions.props
Comment (dougbu) [reply eng/Versions.props:167]: It's worthwhile We discussed this offline and need to invoke the gods: @Pilchie and @ericstj 😺 

We had issues in the past because the updated packages from dotnet/corefx and dotnet/core-setup contained ref/ assemblies with newer assembly versions. This meant the ref/ assemblies built in dotnet/aspnetcore _required_ the newer ref/ assemblies when compiling and that wasn't our intent. @ericstj will newer ref/ assemblies change assembly versions in 5.0 servicing of dotnet/runtime❔

@Pilchie please confirm our requirements here are to
1. include the 5.0.0 ref/ assemblies from dotnet/runtime in our targeting packs
2. include any updated 5.0.x implementation assemblies from dotnet/runtime  in our runtime packs
3. compile against the 5.0.0 ref/ assemblies from dotnet/runtime
4. test against the 5.0.x implementation assemblies from dotnet/runtime

If that's correct, I can try referencing both packages, using `Compile` assets from the 5.0.0 packages and everything else from newer 5.0.x packages. But, I'm doubtful that will work correctly up the project dependency tree in this repo. I'm also not sure how to test this works correctly without dotnet/runtime and us doing a servicing exercise. Suggestions❔

File: eng/targets/ResolveReferences.targets
Comment (dougbu) [reply eng/targets/ResolveReferences.targets:245]: This and the task below were `Warning`s when I added them because I was worried about the Blazor / Mono ref/ assembly project that's since been removed. Will make them `Error`s.

File: eng/Versions.props
Comment (Pilchie) [reply eng/Versions.props:167]: Those requirements seem correct from me.

Comment (wtgodbe) [reply eng/Versions.props:167]: It seems like it should be possible/supported (by roslyn) to always build against the oldest and run against the latest - maybe @agocke or someone else from Roslyn knows how we can accomplish that?

**Review by wtgodbe — CHANGES_REQUESTED** [review]
Everything looks good except for the question about building against old refs & running against new impl's - will approve once that's resolved

File: eng/Versions.props
Comment (Pilchie) [reply eng/Versions.props:167]: That's not a question of the compiler, it's a question of the package contents, and msbuild targets that locate things.  The compiler just gets a path on disk to the thing to reference.

Comment (ericstj) [reply eng/Versions.props:167]: > @ericstj will newer ref/ assemblies change assembly versions in 5.0 servicing of dotnet/runtime

Ref assemblies that are part of the ref-pack will never change in servicing.  Ref assemblies in the transport package we provide for extensions should also not change.

Individual packages will change as needed so that they can have unique assembly versions.

Comment (dougbu) [reply eng/Versions.props:167]: 🆗 @ericstj that means I need to hack things to get our projects compiling against 5.0.0 versions of the assemblies and running against 5.0.x for this specific set of packages. I haven't tried my "reference both versions" idea yet and still have no idea how to test it out without a dotnet/runtime / dotnet/aspnetcore servicing exercise.

Comment (ericstj) [reply eng/Versions.props:167]: Wasn’t this what was already happening in 3.x?  I thought that’s exactly what we made work for the ref pack servicing.  Let me know if there is something different we should be doing to make things easier.

Comment (dougbu) [reply eng/Versions.props:167]: Yes, we're dealing with similar incoming packages in 3.x but aren't using Roslyn to generate the ref/ assemblies. That means it's _much_ easier in 3.x to compile our ref/ assemblies against the 3.1.0 ref/ assemblies from outside aspnetcore. We only need to restrict ourselves to the 3.1.0 versions of these packages when building ref/ projects and Microsoft.AspNetCore.App.Ref. Further, we were already in 3.x servicing when we found the problem which made testing simpler.

With this change, we're building the ref/ and implementation projects with the same `@(Reference)` items and no longer have separate ref/ projects. That makes testing against the 5.0.x packages more complicated and gnarly to test.

Comment (dougbu) [reply eng/Versions.props:167]: Oops, "ref/ and implementation **assemblies**"

File: docs/ReferenceResolution.md
Comment (dougbu) [inline docs/ReferenceResolution.md:48]: Note to self: Make this AspNetCore.sln

File: eng/Versions.props
Comment (dougbu) [inline eng/Versions.props:185]: /FYI this block is about my earlier comment:
> still have no idea how to test it out without a dotnet/runtime / dotnet/aspnetcore servicing exercise.

Idea is to pretend these two packages previous versions are the RTM version and we're at RTM+1. If this works, I'll remove the "test only" commit and let everyone who wants to review know.

File: eng/Versions.props
Comment (dougbu) [inline eng/Versions.props:219]: Change here and at line 255 is about property name consistency. If these properties were automatic, eng/scripts/CodeCheck.ps1 would currently fail; it requires `$(*PackageVersion)` properties for everything in eng/Version.Details.xml. It appears we carried that consistency over to the manual version properties.

If someone (@Pilchie this was your addition) wants to remove `Package` from the names, we can certainly do that -- separately. File an issue and put it on in DoI or Build project. Who knows (@mmitche❔) such a change might even reduce the size of our binlogs or speed builds up a bit 😺

**Comment by ericstj** [comment]
> increases the targeting pack size slightly

One idea we had in dotnet/runtime was to use the ILLinker to trim the resulting reference assemblies that are shipped in the ref pack.  We already do this for our runtime assemblies, doing so for ref might be a good use of linker.  The settings we use with it to root public API and trim unreachable code are here: https://github.com/dotnet/runtime/blob/69b0d160953f5c920e52f021366743623693c158/eng/illink.targets#L192

**Comment by dougbu** [comment]
> One idea we had in dotnet/runtime was to use the ILLinker to trim the resulting reference assemblies that are shipped in the ref pack.

That sounds like fun 😺 but I'm not that sure the small size bump merits the extra work. My latest build produced a 2.02 MB Microsoft.AspNetCore.App.Ref.5.0.0-ci.nupkg while the one from the most recent rolling build of master was 1.74 MB. Thoughts @Pilchie❔

**Comment by Pilchie** [comment]
I suspect it might be a fair bit of work, given that our reference assemblies need to have some internals so that test assemblies can be built that have `InternalsVisibleTo`.  It seems like a significant thing to manage for a negligible benefit.

File: eng/targets/ResolveReferences.targets
Comment (dougbu) [inline eng/targets/ResolveReferences.targets:214]: I believe this is the **Right Thing:tm:** to do but would appreciate confirmation from @Pilchie and @ericstj.

**Review by dougbu — COMMENTED** [review]
🆗 I declare this PR working.

The https://dev.azure.com/dnceng/public/_build/results?buildId=731663 build might still time out (unrelated to my changes) but otherwise did all the **Right Thing:tm:s** Have confirmed the downlevel assemblies landed in Microsoft.AspNetCore.App.Ref.5.0.0-ci.nupkg and shipped assemblies all built against what they should have e.g. Microsoft.AspNetCore.Http.Features.dll references the expected (older) ref/ assembly for System.IO.Pipelines.
``` text
System.Collections, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
System.ComponentModel, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
System.IO.Pipelines, Version=5.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51
System.Linq, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
System.Net.Primitives, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
```

I'll clean up the commits and remove the "test only" parts in preparation for a final review. In the meantime, I would appreciate comments on below `!!!` bit since I'd rather not check that question in 😈

**Comment by dougbu** [comment]
@Pilchie 
> I suspect it might be a fair bit of work

Maybe not quite as much as you suspect. My objection was based on the size of the target @ericstj pointed us to and the potential build perf degradation due to a lot of `<Exec>` invocations. If we did anything, it would be in the Microsoft.AspNetCore.App.Ref project and should not affect anything else building in this repo.

**Comment by ericstj** [comment]
Fair enough.  FWIW we saw a lot larger size differences when we were looking at roslyn's references.  For example, I spot checked System.Text.Json right now and our ref is 25KB vs Roslyn's which is 75KB.  The nice thing about the linker is that it can solve the internals problem for you.  IIRC we were trying idefs and this was hard because the compiler needs legit syntax, whereas with the linker you can root just the internals you need and it will walk the type graph.

**Comment by jaredpar** [comment]
>  FWIW we saw a lot larger size differences when we were looking at roslyn's references. For example, I spot checked System.Text.Json right now and our ref is 25KB vs Roslyn's which is 75KB

This is to some degree expected. The reference assemblies generated by the runtime team are omitting members. This can be done in runtime because they have a good grasp of their specific domain and can safely omit some types / members. Though this has bit them before and ended up causing real world customer issues (the stripping of `struct` fields that were `private` as an example). 

The general compiler feature though must produce reference assemblies that are correct in any customer domain. That means we'll generally include more data because it's a superset of what runtime would produce.

File: eng/targets/ResolveReferences.targets
Comment (dougbu) [reply eng/targets/ResolveReferences.targets:214]: Confirmed offline w/ @Pilchie. We're good to go here.

I'll remove the `!!!` and change the comment to mention the main scenario that would break this target: Layout changes to one of the dozen dotnet/runtime packages could mean we pick a compatible assembly from the new package and the path doesn't exist in the old one. Fortunately, this will break the build with complaints about the invalid strong name. I guess the reverse is also true but can't imagine us _removing_ a TFM from a package during servicing.

**Comment by dougbu** [comment]
:up::date: to consolidate changes since @wtgodbe and @JunTaoLuo reviewed (equivalent of https://github.com/dotnet/aspnetcore/pull/23403/commits/c65704f8782fe331080a17b3f0ea28c5072ae6a8) into two commits, remove the test-only bits, and rebase on preview8 branch. Please review or re-review in case there's anything I missed.

@Pilchie we didn't chat about this but I chose to label this a tell-mode change. LMK if that's not right.

**Comment by dougbu** [comment]
The shift-click menu doesn't always work. The URI for the last 9 commits (all but removing the ref/ projects) is https://github.com/dotnet/aspnetcore/pull/23403/files/32dd707601f5b832060e8bbe49be5197c844545f..7ebf486c9a29c462ad8956d117b224c231111d53 at the moment.

File: eng/RTMVersions/RTMVersions.csproj
Comment (wtgodbe) [inline eng/RTMVersions/RTMVersions.csproj:7]: Can't be parameterized?

**Comment by Pilchie** [comment]
I think tell mode is fine - I'll explicitly mention it in tactics, since it *is* a pretty significant change.

**Review by wtgodbe — APPROVED** [review]

File: eng/RTMVersions/RTMVersions.csproj
Comment (dougbu) [reply eng/RTMVersions/RTMVersions.csproj:7]: Not unless I remove the `Directory.Build.*` files. I'll test out whether that causes any problems after getting this in and doing the merge to 'master'. It's no worse at the moment than the hard-coded TFMs in (say) our project templates. But, I'll clean it up next if possible.

**Comment by Pilchie** [comment]
Yipee!!

**Pull request merged**
